lighthouse-core/test/audits/seo/crawlable-anchors-test.js

```diff
@@ -0,0 +1,102 @@
+ */
+'use strict';
+
+const assert = require('assert');
+const CrawlableAnchorsAudit = require('../../../audits/seo/crawlable-anchors.js');
+
+/* eslint-env jest */
+
+function runAudit({
+  rawHref = '',
+  onclick = '',
+  name = '',
+  hasClickHandler = onclick.trim().length,
+}) {
+  const {score} = CrawlableAnchorsAudit.audit({
+    AnchorElements: [{
+      rawHref,
+      name,
+      hasClickHandler,
+      onclick,
+    }],
+  });
+
+  return score;
+}
+
+describe('SEO: Crawlable anchors audit', () => {
+  it('allows crawlable anchors', () => {
+    assert.equal(runAudit({rawHref: '#top'}), 1, 'hash fragment identifier');
+    assert.equal(runAudit({rawHref: 'mailto:name@example.com'}), 1, 'email link with a mailto URI');
+    assert.equal(runAudit({rawHref: 'https://example.com'}), 1, 'absolute HTTPs URL');
+    assert.equal(runAudit({rawHref: 'foo'}), 1, 'relative URL');
+    assert.equal(runAudit({rawHref: '/foo'}), 1, 'relative URL');
+    assert.equal(runAudit({rawHref: '#:~:text=string'}), 1, 'hyperlink with a text fragment');
+    assert.equal(runAudit({rawHref: 'ftp://myname@host.dom'}), 1, 'an FTP hyperlink');
+    assert.equal(runAudit({rawHref: 'http://172.217.20.78'}), 1, 'IP address based link');
+    assert.equal(runAudit({rawHref: '//example.com'}), 1, 'protocol relative link');
+    assert.equal(runAudit({rawHref: 'tel:5555555'}), 1, 'email link with a tel URI');
+    assert.equal(runAudit({rawHref: '#'}), 1, 'link with only a hash symbol');
+    assert.equal(runAudit({
+      rawHref: '?query=string',
+    }), 1, 'relative link which specifies a query string');
+  });
+
+  it('allows anchors which use a name attribute', () => {
+    assert.equal(runAudit({name: 'name'}), 1, 'link with a name attribute');
+  });
+
+  it('allows anchors which use event listeners on themselves', () => {
+    assert.equal(runAudit({hasClickHandler: true}), 1, 'presence of a click handler is a pass');
+
+    const auditResultJavaScriptURI = runAudit({
+      rawHref: 'javascript:void(0)',
+      hasClickHandler: true,
+    });
+    const assertionMessage = 'hyperlink with a `javascript:` URI and a click handler';
+    assert.equal(auditResultJavaScriptURI, 1, assertionMessage);
+  });
+
+  it('disallows uncrawlable anchors', () => {
+    assert.equal(runAudit({}), 0, 'link with no meaningful attributes and no event handlers');
+    assert.equal(runAudit({rawHref: 'file:///image.png'}), 0, 'hyperlink with a `file:` URI');
+    assert.equal(runAudit({name: ' '}), 0, 'name attribute with only space characters');
+    assert.equal(runAudit({rawHref: ' '}), 0, 'href attribute with only space characters');
+    const assertionMessage = 'onclick attribute with only space characters';
+    assert.equal(runAudit({rawHref: ' ', onclick: ' '}), 0, assertionMessage);
+  });
+
+  it('disallows javascript:void expressions in the onclick attribute', () => {
+    const javaScriptVoidVariations = [
```

Comment:
Happy to turn these each into literal assertions, each on their own line, without looping, if that's preferred.

Reply:
nah this is fine :) it would be nice to throw in a few of the negative cases though to ensure we're not being overly aggressive.

examples cases I had in mind:

`javascript:void0` fail (even though it kinda shouldn't, just a comment documenting it seems fine)
`javascript:void` pass
`windowAlocation` pass
`window.location.href` pass
`window.Location =` pass
`windowLopen()` pass
`window.open('foo', 'name', 'resizable)` fail